cli/cmd/lql_preview.go

```diff
@@ -0,0 +1,129 @@
+
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+
+	"github.com/lacework/go-sdk/api"
+	"github.com/lacework/go-sdk/lwtime"
+)
+
+var (
+	queryPreviewSourceCmd = &cobra.Command{
+		Hidden: false,
+		Use:    "preview-source <datasource_id>",
+		Short:  "Preview Lacework query data source",
+		Long:   `Preview Lacework query data source.`,
+		Args:   cobra.ExactArgs(1),
+		RunE:   previewQuerySource,
+	}
+	queryPreviewSourceTemplate = `CLIAdhocPreview { source { %s } return distinct { %s } }`
+)
+
+func init() {
+	queryCmd.AddCommand(queryPreviewSourceCmd)
+}
+
+func previewQuerySource(_ *cobra.Command, args []string) error {
+	cli.Log.Debugw("retrieving datasource", "id", args[0])
+
+	cli.StartProgress(" Retrieving datasource...")
+	datasourceResponse, err := cli.LwApi.V2.Datasources.Get(args[0])
+	cli.StopProgress()
+
+	if err != nil {
+		return errors.Wrap(err, "unable to retrieve datasource")
+	}
+
+	// build returns list from datasource fields
+	var returns []string
+	for _, ret := range datasourceResponse.Data.ResultSchema {
+		returns = append(returns, ret.Name)
+	}
+	if len(returns) == 0 {
+		return errors.New("unable to parse datasource schema")
+	}
+
+	// initialize time attempts
+	timeAttempts := []map[string]string{
+		map[string]string{"start": "-24h", "end": "now"},
+		map[string]string{"start": "-7d", "end": "-24h"},
+		map[string]string{"start": "-30d", "end": "-7d"},
+	}
+
+	for _, timeAttempt := range timeAttempts {
+		start, _ := lwtime.ParseRelative(timeAttempt["start"])
+		end, _ := lwtime.ParseRelative(timeAttempt["end"])
+
+		queryArgs := []api.ExecuteQueryArgument{
+			api.ExecuteQueryArgument{
+				Name:  "StartTimeRange",
+				Value: start.UTC().Format(lwtime.RFC3339Milli),
+			},
+			api.ExecuteQueryArgument{
+				Name:  "EndTimeRange",
+				Value: end.UTC().Format(lwtime.RFC3339Milli),
+			},
+		}
+
+		// initialize query
+		executeQuery := api.ExecuteQueryRequest{
+			Query: api.ExecuteQuery{
+				QueryText: fmt.Sprintf(
+					queryPreviewSourceTemplate, args[0], strings.Join(returns, ",")),
+			},
+			Arguments: queryArgs,
+		}
+		if args[0] == "CloudTrailRawEvents" {
+			executeQuery.Query.EvaluatorID = "Cloudtrail"
+		}
+
+		// execute query
+		cli.Log.Debugw("running query", "query", executeQuery.Query.QueryText)
+		cli.StartProgress(" Executing preview query...")
+		response, err := cli.LwApi.V2.Query.Execute(executeQuery)
+		cli.StopProgress()
+		if err != nil {
+			return errors.Wrap(err, "unable to preview datasource")
+		}
+
+		// check and output
+		data, ok := response["data"]
+		if !ok {
+			err = errors.New("preview results missing data")
+			return errors.Wrap(err, "unable to preview datasource")
+		}
+		if s := reflect.ValueOf(data); s.Kind() != reflect.Slice {
+			err = errors.New("preview results data is not a slice")
+			return errors.Wrap(err, "unable to preview datasource")
+		}
+		if len(data.([]interface{})) == 0 {
+			continue
+		}
+		return cli.OutputJSON(data.([]interface{})[0])
```

Comment:
Using reflect sounds like too much, we could do the same with a simple "try casting" like this:
```suggestion
		if slice, ok := data.([]interface{}); ok {
			if len(slice) == 0 {
				continue
			}
			return cli.OutputJSON(slice[0])
		} else {
			err = errors.New("preview results data is not a slice")
			return errors.Wrap(err, "unable to preview datasource")
		}
```

Reply:
Fixed.